src/dbcp/transform/acp_projects.py

```diff
@@ -0,0 +1,458 @@
+    )
+    # if lat/lon is present but fails to join to any county, set lat/lon to null UNLESS
+    # the project is an offshore wind project. Those are in the ocean, not in a county.
+    is_offshore = full_df["raw_phase_type"].str.strip() == "Offshore Wind"
+    has_point = ~joined["lat_lon"].is_empty
+    failed_sjoin = joined["GEOID"].isna()
+    joined.loc[
+        failed_sjoin & has_point & ~is_offshore, ["avg_latitude", "avg_longitude"]
+    ] = pd.NA
+    joined.rename(columns={"GEOID": "census_county_id_fips"}, inplace=True)
+
+    # fill in any remaining missing and multi-valued entries with the first value
+    is_missing_multi = joined["county_id_fips"].isna() & is_multivalued
+    first_vals = full_df.loc[is_missing_multi, ["raw_states", "raw_counties"]].apply(
+        lambda s: s.str.split("|", n=1).str[0].str.strip(), axis=1
+    )
+    first_geocoded = (
+        add_county_fips_with_backup_geocoding(
+            first_vals, state_col="raw_states", locality_col="raw_counties"
+        )
+        .rename(
+            columns={
+                "county_id_fips": "geocoded_county_id_fips",
+                "state_id_fips": "geocoded_state_id_fips",
+            }
+        )
+        .convert_dtypes()
+    )
+    geo_cols = [
+        "geocoded_locality_name",
+        "geocoded_locality_type",
+        "geocoded_containing_county",
+        "geocoded_county_id_fips",
+        "geocoded_state_id_fips",
+    ]
+    # fill in new vals
+    joined.loc[:, geo_cols] = joined.loc[:, geo_cols].fillna(first_geocoded[geo_cols])
+    joined["county_id_fips"] = joined["county_id_fips"].fillna(
+        joined["geocoded_county_id_fips"]
+    )
+    # to define state FIPS, use the corresponding county FIPS where it exists. Sometimes
+    # (about 60 cases) only the raw_state is present (no county and either no lat/lon or
+    # lat/lon is in the ocean), so I'll use that too.
+    joined["state_id_fips"] = (
+        joined["county_id_fips"].str[:2].fillna(joined["geocoded_state_id_fips"])
+    )
+    out_cols = [
+        "state_id_fips",
+        "county_id_fips",
+        "avg_latitude",
+        "avg_longitude",
+        "geocoded_locality_name",
+        "geocoded_locality_type",
+        "geocoded_containing_county",
+        "geocoded_county_id_fips",
+        "geocoded_state_id_fips",
+        "census_county_id_fips",
+    ]
+    out = joined[out_cols]
+    return out
+
+
+def _make_surrogate_key(raw_df: pd.DataFrame) -> pd.Series:
+    """Create a surrogate key from several data columns."""
+    # The surrogate key is a hash of several columns that should uniquely identify a row
+    # The main design decisions are whether to use raw data or transformed data as
+    # inputs, and how many columns to include in the hash.
+    # The transformed data risks changes in the transform functions that break backwards
+    # compatibility. The raw data risks changes in the raw data that break backwards
```

Comment:
I think hashing some columns to make a surrogate key is a good idea! We should probably do this for the lbnl data too so we have steady ids throughout the our warehouse and data marts.